packages/next/client/use-intersection.tsx

```diff
@@ -34,26 +34,25 @@ export function useIntersection<T extends Element>({
 
   const unobserve = useRef<Function>()
   const [visible, setVisible] = useState(false)
-  const [root, setRoot] = useState(rootRef ? rootRef.current : null)
-  const setRef = useCallback(
-    (el: T | null) => {
-      if (unobserve.current) {
-        unobserve.current()
-        unobserve.current = undefined
-      }
+  const elementRef = useRef<T | null>(null)
 
-      if (isDisabled || visible) return
+  useEffect(() => {
+    if (unobserve.current) {
+      unobserve.current()
+      unobserve.current = undefined
+    }
 
-      if (el && el.tagName) {
-        unobserve.current = observe(
-          el,
-          (isVisible) => isVisible && setVisible(isVisible),
-          { root, rootMargin }
-        )
-      }
-    },
-    [isDisabled, root, rootMargin, visible]
-  )
+    if (isDisabled || visible) return
+
+    const el = elementRef.current
+    if (el && el.tagName) {
+      unobserve.current = observe(
+        el,
+        (isVisible) => isVisible && setVisible(isVisible),
+        { root: rootRef?.current, rootMargin }
+      )
+    }
+  }, [isDisabled, rootMargin, rootRef, visible])
```

Comment:
**Before the PR**

- `setIntersection` is used in `<Image />`'s callback ref, which will be called when `<Image />` is first mounted.
- During `setIntersection`, an IntersectionObserver instance is created and it observes the `HTMLImageElement` with `rootRef.current` is still `null`
- By the time `<Image />` is mounted, its parent element (`lazyRoot`) hasn't been mounted yet, so `lazyRoot` has a value of `{ current: null }`.
- `useIntersection` - `rootRef` remains the same identity, thus the `setRoot` effect will never be called again.
- And the `lazyRoot` is never applied to the IntersectionObserver instance

**After the PR**

- `setIntersection` is used in `<Image />`'s callback ref, which will be called when `<Image />` is first mounted.
- During `setIntersection`, the HTMLImageElement is synced to a ref (use ref instead of state to prevent extra re-rendering)
- By the time `useEffect` commits the effect, `lazyRoot` is already mounted, and the `rootRef.current` exists
- Only then the observer is created with `lazyRoot` applied.